squint.c

```diff
@@ -0,0 +1,1611 @@
+int *relocate_nop(int *funcBegin, int *funcEnd, int mode)
+{
+   int *retVal = funcEnd;
+   int *memblk;
+   int *branchAddr;
+   int *branchTarget;
+   int *permutation;
+   int ii, jj, tmp, done;
+   int currAddr;
+   int currTarget;
+   int hasLink;
+   int offset;
+   int align;
+
+   int *scan;
+   int *packed;
+
+   /* count number of branches to relocate */
+   int nopCount = 0;
+   int branchCount = 0;
+
+   for (scan = funcBegin; scan < funcEnd; ++scan) {
+      while (is_const(scan)) ++scan;
+
+      // legal since rename_nop occurs before this call...
+      if (*scan == NOP) {
+         ++nopCount;
+      }
+      else if ((*scan & 0x0e000000) == 0x0a000000) {
+         hasLink = *scan & (1<<24);
+         if ( ((mode == 0) && !hasLink) ||   // intra-function mode
+              ((mode != 0) &&  hasLink) ) {  // inter-function mode
+            ++branchCount;
+         }
+      }
+   }
+
+   if (branchCount != 0) {
+      memblk = malloc(3*branchCount*4);
+      branchAddr = memblk;
+      branchTarget = branchAddr + branchCount;
+      permutation = branchTarget + branchCount;
+
+      /* record all branch inst and target addresses */
+      branchCount = 0;
+      for (scan = funcBegin; scan < funcEnd; ++scan) {
+         while (is_const(scan)) ++scan;
+
+         if ((*scan & 0x0e000000) == 0x0a000000) {
+            hasLink = *scan & (1<<24);
+            if ( ((mode == 0) && !hasLink) ||   // intra-function mode
+                 ((mode != 0) &&  hasLink) ) {  // inter-function mode
+               /* add branch to table */
+               permutation[branchCount] = branchCount;
+               branchAddr[branchCount] = (scan - funcBegin);
+               tmp = (*scan & 0x00ffffff) |
+                     ((*scan & 0x00800000) ? 0xff000000 : 0);
+               branchTarget[branchCount] = (scan - funcBegin) + 2 + tmp;
+               ++branchCount;
+            }
+         }
+      }
+
+      /* sort in terms of target address */
+      /* this is an order preserving sort */
+      /* for secondary keys */
+      ii = 0;
+      do {
+         done = 1;
+         for (jj = branchCount-2; jj >= ii; --jj) {
+            if (branchTarget[jj+1] < branchTarget[jj]) {
+               tmp = branchTarget[jj+1];
+               branchTarget[jj+1] = branchTarget[jj];
+               branchTarget[jj] = tmp;
+               tmp = permutation[jj+1];
+               permutation[jj+1] = permutation[jj];
+               permutation[jj] = tmp;
+               done = 0;
+            }
+         }
+         ++ii;
+      } while (!done);
+   }
+
+   if (nopCount > 0) { // there are nops to remove
+
+      currAddr = 0;
+      currTarget = 0;
+      offset = 0;
+
+      packed = funcBegin;
+      for (scan = funcBegin; scan <= funcEnd; ++scan, ++offset) {
+         align = 1;
+         while (currTarget < branchCount) {
+            if (branchTarget[currTarget] == offset) {
+               if (align && ((mode != 0) ||
```

Comment:
From the A72 optimization guide:

• Consider aligning subroutine entry points and branch targets to quadword boundaries, within the bounds of the code-density requirements of the program. This will ensure that the subsequent fetch can retrieve four (or a full quadword’s worth of) instructions, maximizing fetch bandwidth following the taken branch.

I'm assuming that a jump flushes the pipeline, and the optimization guide says alignment will maximally refill. I trust the optimization guide.  I only align loops and function calls.